products/limesurvey.md

```diff
@@ -0,0 +1,166 @@
+    eol: 2022-02-14
+    latest: "5.2.14"
+    latestReleaseDate: 2022-02-14
+-   releaseCycle: "5.1"
+    releaseDate: 2021-08-11
+    support: 2021-11-10
+    eol: 2021-11-10
+    latest: "5.1.18"
+    latestReleaseDate: 2021-11-01
+-   releaseCycle: "5.0"
+    releaseDate: 2021-05-26
+    support: 2021-08-11
+    eol: 2021-08-11
+    latest: "5.0.13"
+    latestReleaseDate: 2021-08-04
+
+-   releaseCycle: "4.6"
+    releaseDate: 2021-05-04
+    support: false
+    eol: true
+    latest: "4.6.3"
+    latestReleaseDate: 2021-05-18
+-   releaseCycle: "4.5"
+    releaseDate: 2021-04-12
+    support: 2021-05-04
+    eol: 2021-05-04
+    latest: "4.5.2"
+    latestReleaseDate: 2021-04-26
+-   releaseCycle: "4.4"
+    releaseDate: 2021-01-29
+    support: 2021-04-12
+    eol: 2021-04-12
+    latest: "4.4.16"
+    latestReleaseDate: 2021-04-06
+-   releaseCycle: "4.3"
+    releaseDate: 2020-06-16
+    support: 2021-01-29
+    eol: 2021-01-29
+    latest: "4.3.34"
+    latestReleaseDate: 2021-01-19
+-   releaseCycle: "4.2"
+    releaseDate: 2020-04-22
+    support: 2020-06-16
+    eol: 2020-06-16
+    latest: "4.2.8"
+    latestReleaseDate: 2020-06-08
+-   releaseCycle: "4.1"
+    releaseDate: 2020-01-28
+    support: 2020-04-22
+    eol: 2020-04-22
+    latest: "4.1.18"
+    latestReleaseDate: 2020-04-16
+-   releaseCycle: "4.0"
+    releaseDate: 2020-01-16
+    support: 2020-01-28
+    eol: 2020-01-28
+    latest: "4.0.1"
+    latestReleaseDate: 2020-01-20
+
+-   releaseCycle: "3.28"
+    releaseDate: 2022-02-14
+    support: false
+    eol: 2023-07-31
+    latest: "3.28.64"
+    latestReleaseDate: 2023-07-05
+-   releaseCycle: "3.27"
+    releaseDate: 2021-05-25
+    support: 2022-02-14
+    eol: 2022-02-14
+    latest: "3.27.35"
+    latestReleaseDate: 2022-02-08
+-   releaseCycle: "3.26"
+    releaseDate: 2021-04-19
+    support: 2021-05-25
+    eol: 2021-05-25
+    latest: "3.26.5"
+    latestReleaseDate: 2021-05-19
+-   releaseCycle: "3.25"
+    releaseDate: 2020-11-17
+    support: 2021-04-19
+    eol: 2021-04-19
+    latest: "3.25.22"
+    latestReleaseDate: 2021-04-13
+-   releaseCycle: "3.24"
+    releaseDate: 2020-10-13
+    support: 2020-11-17
+    eol: 2020-11-17
+    latest: "3.24.6"
+    latestReleaseDate: 2020-11-09
+-   releaseCycle: "3.23"
+    releaseDate: 2020-08-13
+    support: 2020-10-13
+    eol: 2020-10-13
+    latest: "3.23.7"
+    latestReleaseDate: 2020-10-06
+
+---
+
+> [LimeSurvey](https://www.limesurvey.org/) is a free and open source online statistical survey web app written in PHP based on a relational database, distributed under the GNU General Public License.
+
+LimeSurvey usually supports multiple major versions at a time, however, within a major version support is limited to the respective latest version. The support of minor versions ends by the release of the next minor version (i.e. the release of LimeSurvey 6.1 ends the support of LimeSurvey 6.0). Each major version will have support periods as follows:
```

Comment:
Given only the latest minor is supported inside a major, I think it makes more sense to set the cycle on the major version instead of the minor. With a fixed EOL date it looks odd to have to update the EOL date each time there is a new minor release.

This way there is no need for a support column (support fields can be removed too) and it will make things easier to maintain over time.

WDYT ?

Reply:
Aaaaargh, GitHub, damn you! I just wrote a very long answer, figured I wanted to commit first and the page reloaded - and everything was gone... :cursing_face: I'll rewrite it, but later...

Reply:
I totally agree content-wise: The "true" release cycle is LimeSurvey 6, not LimeSurvey 6.1, because support of LimeSurvey 6.1 will end as soon as LimeSurvey 6.2 is released.

However, IMHO we should keep in mind what a user might expect from `endoflife.date` considering LimeSurvey. What `endoflife.date` considers the release cycle doesn't have to match the "true" release cycle, we should rather take a practical stance here. Due to how `endoflife.date` presents its data, one can only access a release cycle's information - both on the website, and via the API. If we take the major version as release cycle all information between major versions is lost - even when such cycles last multiple years and consist of many minor versions. From a LimeSurvey user's perspective one rather wants to know when the support of a minor version has ended, simply because this happens way more frequently.

Due to how `endoflife.date` currently works, we'd be unable to do that. We can't request the EOL date of LimeSurvey 6.0, we can't even use the API to check whether LimeSurvey 6.0 is still supported. We'd have to check ourselves whether the latest version matches our minor version - i.e. we'd have to reimplement parts of `endoflife.date`'s logic. I don't think that this is a good thing. Some other projects here on `endoflife.date` do the same (e.g. phpMyAdmin), even though their "true" release cycle is different. Major versions are rare - simply because mature projects don't make drastic changes and/or break stuff, people rather create new projects if they want to do things differently. And, in terms of semantic versioning, because there's often some debate on what a breaking change is... :wink: I just believe that this is what users would expect for LimeSurvey - and from many other projects too (e.g. phpMyAdmin).

Thus I'd vote for using the minor versions as release cycles, even though that's not strictly true.

However, not for ever. I noticed that you're currently working on a new API (#2080) and improved automation. I didn't check the code, only saw the API preview - and it looks very promising. Great work! :+1: IMHO the most important change besides the new API design should be removing the segregation of release cycles. Considering `endoflife-date/release-data` I don't really see a reason why one shouldn't be able to request the support status of arbitrary versions, not just release cycles. For example, for versions more specific than the project's release cycle (e.g. version 1.2.3), `endoflife.date` can and should assume that any but the latest version is supported unless declared otherwise (i.e. it's a release cycle). The EOL date can then be taken from the respective next version's release date. For versions less specific than the project's release cycle (e.g. version 1 if minor versions are the supposed release cycle), `endoflife.data` should simply return the data of the matching latest release cycle. This doesn't just heavily improve what one can do with `endoflife.date`, it even aids automation: Since adding a new release cycle currently requires manual intervention, projects like LimeSurvey can safely switch to major versions as release cycle and therefore require such manual intervention way less frequently.

Reply:
I understand your point about not being able to know when a minor version ended, but we could say the same for patch versions regarding this matter.

You could argue that we already took the same approach for products like https://endoflife.date/quarkus-framework, https://endoflife.date/phpmyadmin or https://endoflife.date/keycloak but there are some differences:

- those product do not have predictable EOL dates, meaning the rule for those is `eol(x) = releaseDate(x+1)`,
- we want to document extended support (aka commercial support) for those specific cycles,
- or there are multiple level of support always for those specific cycles.

Being able to request the support status of arbitrary versions from the API would be a great thing to support in the future. But it will not be easy to do: like you said endoflife.date is very 'cycle-oriented', and the API is not dynamic (it's just a bunch of generated JSON files), so the rule you describe must be computed on client's side.

I am raising the point because, as far as I know, there is no other product on endoflife.date for which we documented minor release cycle without having at least one of the reason above. I am not against documenting minor versions, but would like to have some feedback from the team before. @endoflife-date/everyone what do you think ?

Reply:
> I understand your point about not being able to know when a minor version ended, but we could say the same for patch versions regarding this matter.

Yes, I agree, that's why I suggested to allow that in the future :wink: The problem is that `endoflife.date` currently only can provide information about either major versions, minor versions, or patch versions - i.e. about release cycles. Unless `endoflife.date` removes that limitation we should choose what most users expect - and for LimeSurvey that's IMHO minor versions.

> You could argue that we already took the same approach for products like https://endoflife.date/quarkus-framework, https://endoflife.date/phpmyadmin or https://endoflife.date/keycloak but there are some differences:

Sounds reasonable. However, one does not exclude the other? All these reasons require `endoflife.date` to use minor versions instead of major versions, but are there any reasons why we can't use minor versions here? Yes, major versions are the "true" release cycles for LimeSurvey, but since that imposes some disadvantages, what are the advantages to counter those?

Since you're working on a new API right now it's IMHO the perfect time to think about things like that. I'm trying to provide an outsider's perspective on this. Abandoning the strict release-cycle-orientation and rather working on versions instead would IMHO make `endoflife.date` way more powerful. The release cycle still remains the information one needs to provide in the Markdown file, but from the API's perspective, release cycles should no longer dictate the information provided, but are rather just used for what they are actually for: To determine the support status of versions. Nothing changes for products without automation (i.e. if we have no release data).

Definitely looking forward to more opinions on this! :heart: Maybe we should open a separate issue? A PR adding a single product probably isn't the best place for such a general discussion.

> Being able to request the support status of arbitrary versions from the API would be a great thing to support in the future. But it will not be easy to do: like you said endoflife.date is very 'cycle-oriented', and the API is not dynamic (it's just a bunch of generated JSON files), so the rule you describe must be computed on client's side.

By "arbitrary versions" I don't mean non-existing versions like "LimeSurvey 13.37", but all known versions, e.g. LimeSurvey 6.0.0, 6.0.1, 6.0.2, … Basically I'm suggesting that `endoflife.date` should add API responses for every version in `endoflife-date/release-data`, as well as the respective version groups, no matter whether it's a release cycle or not (i.e. for version `1.2.3` we add responses for `1.2.3`, `1.2`, and `1`).

I'm aware that `endoflife.date` uses Jekyll. I know Jekyll. As a static website generator we can't have dynamic responses, but do we need dynamic responses? We know the versions after all. I'm not aware of any limitation that would prevent us from creating a JSON file for every single version, i.e. `1.2.0.json`, `1.2.1.json`, `1.2.2.json`, …, `1.2.json`, `1.json`, … Yes, a lot of files, but does that matter? No client-side computations necessary.

Important: I'm not suggesting removing the `/products/{product}/cycles/` endpoint. I rather suggest adding a new one like `/products/{product}/versions/` (i.e. `/products/{product}/versions/1.2.3`, `/products/{product}/versions/1.2`, `/products/{product}/versions/1`). For versions matching the release cycle it might return the same JSON. Another advantage of this: Users don't have to know what the release cycle is.

Reply:
> If we take the major version as release cycle all information between major versions is lost - even when such cycles last multiple years and consist of many minor versions. From a LimeSurvey user's perspective one rather wants to know when the support of a minor version has ended, simply because this happens way more frequently.

So the relevant information that might be useful (and go missing) to users here is:

1. Release date of each minor release
2. Support/EOL lifecycles of each minor release.

In general, we always try to collapse rows whenever we can, to reduce maintainer burden. Unless there is clearly some differentiating feature between rows (I haven't thought through all the criteria, but the list @marcwrobel posted above seemed reasonable).

In general, I always treat our guidance to favor "What version should I be running", instead of keeping track of every data point about the project (that can grow into a lot). And collapsing rows still keeps this answerable easily, imo, even without the two pieces of information that go "missing".

I'd vote to collapse rows, mainly on the maintenance burden, and harmonization for now (we need our pages to follow similar guidance).

> Due to how endoflife.date currently works, we'd be unable to do that. We can't request the EOL date of LimeSurvey 6.0, we can't even use the API to check whether LimeSurvey 6.0 is still supported. We'd have to check ourselves whether the latest version matches our minor version - i.e. we'd have to reimplement parts of endoflife.date's logic. I don't think that this is a good thing.

I agree with this gripe. Our API should support versions, and be easy to integrate. Hence my desire to keep a ["supportedVersions/supportedReleases" in our specification](https://github.com/endoflife-date/releases.json/pull/1/files#r1052509408) to make it easy for clients to decide whether they are on a supported release (Do a "version in supportedVersions" check instead of worrying about cycles).

We can easily support some of the data you want here (exact EOL dates for each of the _patch version_ even), since we have a list of all release dates via the automation.

Within a release cycle, go through all patch release, and "generate an EOL date" as soon as the next patch release comes out.

So even if we track things on a "major" basis, we can calculate:

```prolog
eol(6.0.0) = releaseDate(6.0.1)
eol(6.0.1) = releaseDate(6.0.2)
...
eol(6.0.7) = releaseDate(6.1.0)
...
eol(6.1.5) = releaseDate(6.1.6)
eol(6.1.6) = eol(6.1) # Cycle EOL
```

And provide this information in the API quite easily. I can think of a "splitCycles: true" attribute as well, which will do the same for the website as well.

with just a single row:
```yaml
-   releaseCycle: "6"
    releaseDate: 2023-04-05 # autogenerated
    support: 2025-04-30
    eol: 2026-04-30
    latest: "6.1.6" # autogenereated
    latestReleaseDate: 2023-07-03 #autogenerated
```

At the end, the way we display the information (website/API) is where we care about usability (for both end-users and programmatic usage). For the source information itself, it is more important that we focus on maintainers and maintainability instead, we can always display things differently, as long as we have this information (which we will, in this case).

Reply:
@PhrozenByte are you OK with dropping minor releases ? Do you want me to take care of it ?

Reply:
> @PhrozenByte are you OK with dropping minor releases ?

Not unless a decision is made towards the API. If the API is updated to support versions: Yes. We can drop minor versions temporarily to get this merged then, simply because we can be sure that the issue will be resolved in the future. However, if not, the answer is No.

I'm afraid that LimeSurvey users just don't expect major versions here. If the API is updated to allow querying for minor versions too, then there's no issue and I'm absolutely fine with dropping minor releases in the Markdown file. However, if the API isn't going to support versions, I'm afraid that we more or less yield the API for LimeSurvey nearly worthless.

The reason is simple: LimeSurvey does not have predictable EOL dates. I'm pretty much following @marcwrobel's criteria here :wink: On a side note, the existence of LimeSurvey 3.28 LTS and its unpredictable EOL date (it's not `releaseDate(x+1)`) further reinforces this point.

LimeSurvey's minor versions don't have a predictable EOL date, because there's no release schedule for minor releases. Any next version can be a minor release, yielding the previous minor version EOL. This is very problematic in practice, because apart from LimeSurvey's paid ComfortUpdate feature, one must perform updates manually. For patch releases this is no big issue, you simply replace the source files - something one can easily automate. However, this is not possible for minor releases: A minor version might change the database scheme, requiring manual interference. Thus one must know when a minor release reached its EOL to run a manual update, just as with major releases. That's where `endoflife.date` comes into play: With the API one can check whether a minor version reached its EOL and notify the user accordingly. When dropping minor versions we can't do that.

Testing whether a minor release reached its EOL is - by far - the most important use case of `endoflife.date`'s API for LimeSurvey users. I'm aware that this might be different for most other projects, but most other projects don't have paid auto updaters either.

Reply:
> Not unless a decision is made towards the API. If the API is updated to support versions: Yes.

I would be happy to review any PR that improve the API in this direction.

> The reason is simple: LimeSurvey does not have predictable EOL dates.

I don't agree with this statement : https://manual.limesurvey.org/LimeSurvey_roadmap document EOL dates for their major releases. It may be updated along the way, but that's not something uncommon and we can update those dates manually if it ever happen. As you said it, the unpredictability is happening when using minor releases.

> Testing whether a minor release reached its EOL is - by far - the most important use case of endoflife.date's API for LimeSurvey users.

That can be said for patch releases too. That's why I agree that a new endpoint allowing to check if a given version is supported or not would be a really nice addition to the API.

And that's also already possible with the current API with a little bit more client-side computation:

- extracting the cycle from the version to test,
- fetching the cycle EOL date (or the latest release) from the API (depending on the language it is more or less simple, example with bash / httpie / jq : `http https://deploy-preview-3266--endoflife-date.netlify.app/api/limesurvey.json | jq '.[] | select(.cycle == "3.25") | .eol'`),
- and comparing that EOL date to the current date (or the latest release to the current release).

Reply:
What should we do @endoflife-date/everyone ? Use major releases (which is more in line with how endoflife.date works currently and is easier to maintain) or minor releases (which _may_ be more practical for API users but a little bit harder to maintain) ?

Reply:
@PhrozenByte Would you be willing to track releases for LimeSurvey and take ownership of the page for updates on new releases?

We could use minor releases for now in that case, till we get the APIs figured out.

Reply:
> @PhrozenByte Would you be willing to track releases for LimeSurvey and take ownership of the page for updates on new releases?

Yes. However, I'd need #3311 first. The problem I'm trying to solve with `endoflife.date` is to learn about new minor releases, so taking responsibility without a way to get notified about new releases is kind of a chicken-and-egg problem :laughing: However, #3311 adds a maintainers / ownership concept to products, so that Dependabot could automatically notify me when a new / unknown release was found, allowing me to add it manually then.

> I would be happy to review any PR that improve the API in this direction.

Following #2080 you're using a Jekyll plugin to create the JSON files, correct? I've have never written a single line of Ruby code in my life, so I'm afraid that I can't provide any help on this matter.

As I said, I don't expect immediate implementation, as mentioned earlier I'm fine with either temporarily dropping minor releases, or taking responsibility for updates in the meantime, however, I must know that this will make it into the v1 API. More feedback about the idea is very welcome so that we can improve the idea and the project can make a decision about whether it's coming or not.

Shall I add a comment to #2080 to ask more people for feedback about the suggested API endpoint?

> That can be said for patch releases too.

As I was trying to explain earlier, new patch releases are no problem for LimeSurvey, because they never require manual interference. Minor releases do require manual interference - thus it's crucial to learn about the EOL date of minor releases. I didn't open this PR as an end in itself, I want to actually use it.

Anyway and just for the record: The suggested API changes would fix this for patch releases, too.